File: recipes/subms-events-store/python/src/subms_events_store/event_store.py
"""The append-only event log. Stores subms-events Events, returns an offset per
append, and fans appended events to live subscribers via an EventDispatcher."""

from __future__ import annotations

from typing import Iterator, List, Optional

from subms_events import DispatchMode, Event, EventDispatcher, EventListener


class EventStore:
    """In-memory, append-only log with offset addressing + live subscriptions.
    Durability is out of scope - pair with subms-ts-wal to persist."""

    def __init__(self, dispatch: DispatchMode = DispatchMode.SYNC) -> None:
        self._log: List[Event] = []
        self._dispatcher = EventDispatcher(dispatch)

    def append(self, event: Event) -> int:
        offset = len(self._log)
        self._log.append(event)
        self._dispatcher.emit(event)
        return offset

    def __len__(self) -> int:
        return len(self._log)

    def is_empty(self) -> bool:
        return not self._log

    def get(self, offset: int) -> Optional[Event]:
        if 0 <= offset < len(self._log):
            return self._log[offset]
        return None

    def events(self) -> List[Event]:
        return self._log

    def read_from(self, offset: int) -> List[Event]:
        i = min(offset, len(self._log))
        return self._log[i:]

    def by_topic(self, topic: str) -> Iterator[Event]:
        return (e for e in self._log if e.topic == topic)

    def subscribe(self, listener: EventListener) -> None:
        self._dispatcher.add_listener(listener)

    def stop(self) -> None:
        self._dispatcher.stop()

    def to_json(self) -> str:
        return "[" + ",".join(e.to_json() for e in self._log) + "]"

File: recipes/subms-events-store/rust/src/event_store.rs
//! The append-only event log. Stores `subms-events` `Event`s, hands back an
//! offset per append, and (optionally) fans each appended event to live
//! subscribers through an `EventDispatcher`.

use std::sync::Arc;

use subms_events::{DispatchMode, Event, EventDispatcher, EventListener};

/// An in-memory, append-only log of events with offset addressing and live
/// subscriptions. Durability is out of scope - pair with `subms-ts-wal` to
/// persist the log.
pub struct EventStore {
    log: Vec<Event>,
    dispatcher: EventDispatcher,
}

impl EventStore {
    /// A store whose subscribers run inline on `append` (sync dispatch).
    pub fn new() -> Self {
        Self {
            log: Vec::new(),
            dispatcher: EventDispatcher::sync(),
        }
    }

    /// Choose how subscribers are notified (sync inline / async off-thread).
    pub fn with_dispatch(mode: DispatchMode) -> Self {
        Self {
            log: Vec::new(),
            dispatcher: EventDispatcher::new(mode),
        }
    }

    /// Append an event; returns its 0-based offset. Subscribers are notified.
    pub fn append(&mut self, event: Event) -> u64 {
        let offset = self.log.len() as u64;
        self.log.push(event.clone());
        self.dispatcher.emit(event);
        offset
    }

    pub fn len(&self) -> usize {
        self.log.len()
    }

    pub fn is_empty(&self) -> bool {
        self.log.is_empty()
    }

    /// The event at `offset`, if any.
    pub fn get(&self, offset: u64) -> Option<&Event> {
        self.log.get(offset as usize)
    }

    /// The whole log.
    pub fn events(&self) -> &[Event] {
        &self.log
    }

    /// Events at `offset..` (empty if `offset` is past the end).
    pub fn read_from(&self, offset: u64) -> &[Event] {
        let i = (offset as usize).min(self.log.len());
        &self.log[i..]
    }

    /// Events whose topic equals `topic`, in order.
    pub fn by_topic<'a>(&'a self, topic: &'a str) -> impl Iterator<Item = &'a Event> + 'a {
        self.log.iter().filter(move |e| e.topic == topic)
    }

    /// Register a live subscriber. It receives every event appended after this
    /// call (and, in async mode, on the dispatcher thread).
    pub fn subscribe(&mut self, listener: Arc<dyn EventListener>) {
        self.dispatcher.add_listener(listener);
    }

    /// Serialize the log as a JSON array of events. Deterministic and
    /// byte-equivalent across the language ports.
    pub fn to_json(&self) -> String {
        let mut out = String::from("[");
        for (i, e) in self.log.iter().enumerate() {
            if i > 0 {
                out.push(',');
            }
            out.push_str(&e.to_json());
        }
        out.push(']');
        out
    }
}

impl Default for EventStore {
    fn default() -> Self {
        Self::new()
    }
}
